modules/querier/external/client.go

```diff
@@ -0,0 +1,197 @@
+package external
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"time"
+
+	"github.com/cristalhq/hedgedhttp"
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/grafana/tempo/pkg/api"
+	"github.com/grafana/tempo/pkg/hedgedmetrics"
+	"github.com/grafana/tempo/pkg/tempopb"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/weaveworks/common/user"
+	"golang.org/x/oauth2"
+)
+
+var (
+	metricEndpointDuration = promauto.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: "tempo",
+		Name:      "querier_external_endpoint_duration_seconds",
+		Help:      "The duration of the external endpoints.",
+		Buckets:   prometheus.DefBuckets,
+	}, []string{"endpoint"})
+	metricExternalHedgedRequests = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Namespace: "tempo",
+			Name:      "querier_external_endpoint_hedged_roundtrips_total",
+			Help:      "Total number of hedged external requests. Registered as a gauge for code sanity. This is a counter.",
+		},
+	)
+)
+
+type Config struct {
+	Backend string
+
+	HTTPConfig     *HTTPConfig
+	CloudRunConfig *CloudRunConfig
+}
+
+type Client struct {
+	httpClient    *http.Client
+	endpoints     []string
+	tokenProvider tokenProvider
+}
+
+type CloudRunConfig struct {
+	Endpoints         []string      `yaml:"external_endpoints"`
+	HedgeRequestsAt   time.Duration `yaml:"external_hedge_requests_at"`
+	HedgeRequestsUpTo int           `yaml:"external_hedge_requests_up_to"`
+}
+
+type HTTPConfig struct {
+	Endpoints         []string
+	HedgeRequestsAt   time.Duration
+	HedgeRequestsUpTo int
+}
+
+type tokenProvider interface {
+	// Returns an oauth2 token, leveraging a cache unless the token is expired.
+	// If expired, the token is renewed and added to the cache.
+	//
+	// If this returns nil, the request will be unauthenticated.
+	getToken(ctx context.Context, endpoint string) (*oauth2.Token, error)
+}
+
+type option func(client *Client) error
+
+func withTokenProvider(provider tokenProvider) option {
+	return func(client *Client) error {
+		client.tokenProvider = provider
+		return nil
+	}
+}
+
+func NewClient(cfg *Config) (*Client, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	switch cfg.Backend {
+	case "":
+		// For backwards compatibility, use unauthenticated http as the default.
+		return newClientWithOpts(&commonConfig{
+			endpoints:         cfg.HTTPConfig.Endpoints,
+			hedgeRequestsAt:   cfg.HTTPConfig.HedgeRequestsAt,
+			hedgeRequestsUpTo: cfg.HTTPConfig.HedgeRequestsUpTo,
+		})
+	case "google_cloud_run":
+		provider, err := newGoogleProvider(ctx, cfg.CloudRunConfig.Endpoints)
+		if err != nil {
+			return nil, err
+		}
+		return newClientWithOpts(&commonConfig{
```

Comment:
i expect this to eventually drift into two different clients, but i think this is good for now.